[peer] Refactor peer code into its own package

tuxcanfly opened the pull request:
This pull request introduces package peer which contains peer related features
refactored from `peer.go`.

The following is an overview of the features the package provides:

- Provides a basic concurrent safe bitcoin peer for handling bitcoin
    communications via the peer-to-peer protocol
- Full duplex reading and writing of bitcoin protocol messages
- Automatic handling of the initial handshake process including protocol
    version negotiation
- Automatic periodic keep-alive pinging and pong responses
- Asynchronous message queuing of outbound messages with optional
    channel for notification when the message is actually sent
- Inventory message batching and send trickling with known inventory
    detection and avoidance
- Ability to wait for shutdown/disconnect
- Flexible peer configuration
    - Caller is responsible for creating outgoing connections and listening
    for incoming connections so they have flexibility to establish
    connections as they see fit (proxies, etc.)
    - User agent name and version
    - Bitcoin network
    - Service support signalling (full nodes, bloom filters, etc.)
    - Maximum supported protocol version
    - Ability to register callbacks for handling bitcoin protocol messages
- Proper handling of bloom filter related commands when the caller does
    not specify the related flag to signal support
    - Disconnects the peer when the protocol version is high enough
    - Does not invoke the related callbacks for older protocol versions
- Snapshottable peer statistics such as the total number of bytes read
    and written, the remote address, user agent, and negotiated protocol
    version
- Helper functions for pushing `addresses`, `getblocks`, `getheaders`, and
    `reject` messages
    - These could all be sent manually via the standard message output
    function, but the helpers provide additional nice functionality such
    as duplicate filtering and address randomization
- Full documentation with example usage
- Test coverage

In addition to the addition of the new package, `btcd` has been refactored
to make use of the new package by extending the basic peer it provides to
work with the `blockmanager` and `server` to act as a full node.  The
following is a broad overview of the changes to integrate the package:

- The `server` is responsible for all connection management including
    persistent peers and banning
- Callbacks for all messages that are required to implement a full node
    are registered
- Logic necessary to serve data and behave as a full node is now in the
    callback registered with the peer

Finally, the following peer-related things have been improved as a part
of this refactor:

- Don't log or send reject message due to peer disconnects
- Remove trace logs that aren't particularly helpful
- Finish an old TODO to switch the queue `WaitGroup` over to a channel
- Improve various comments and fix some code consistency cases
- Improve a few logging bits
- Implement a most-recently-used nonce tracking for detecting self
    connections and generate a unique nonce for each peer

tuxcanfly commented:
Update: Moved all message handlers into respective listeners. Listeners can be added and removed using a key (string) and all listeners are invoked when a message is received.

Looking into a couple of issues which cropped up while testing:

- [ ] hang on btcctl peer commands
- [ ] no pong replies to ping messages
- [ ] hang on shutdown on outbound peer

davecgh commented:
I'll be giving this a thorough review over the coming days, but a few points from a very quick review are:

- I think there is too much state related to the block manager left in the peer (requested transactions, requested blocks, etc).  Keep in mind the point of having this peer code separated is common code for full nodes, SPV nodes, proxies, etc.  State related to the full node's block manager needs to be in the block manager, not the peer.
- The semantic version bits don't belong in the peer code.  That is something defined by the caller, not the peer.
- The callback registration bits should be defined at the file level as functions rather than function pointers to closures.
- All of the callback registration functions need comments
- The `peer/log.go` is completely wrong.  It should be like `blockchain/log.go`, not have all of the btcd subsystems.

davecgh commented on blockmanager.go:
```diff
@@ -1273,10 +1280,10 @@ func (b *blockManager) NewPeer(p *peer) {
 	// Don't accept more transactions if we're shutting down.
 	if atomic.LoadInt32(&b.shutdown) != 0 {
-		p.txProcessed <- struct{}{}
+		p.TxProcessed <- struct{}{}
```
I believe we're going to want a better mechanism for this.  It is mixing peer and block manager state and won't fit in well with other callers.

davecgh commented on blockmanager.go:
```diff
@@ -1284,10 +1291,10 @@ func (b *blockManager) QueueTx(tx *btcutil.Tx, p *peer) {
 	// Don't accept more blocks if we're shutting down.
 	if atomic.LoadInt32(&b.shutdown) != 0 {
-		p.blockProcessed <- struct{}{}
+		p.BlockProcessed <- struct{}{}
```
Same here about mixing states.

davecgh commented:
I see no reason to register the listeners with a key.  Just use the function itself since each one will have a unique address.

I'd also make the listeners registration data structure a slice over a map since they are more efficient for the general case.  Note that this change would mean you need to walk the slice to find which entry to delete when removing a listener, but it is highly unlikely there are going to be a large number of listeners anyways, and it's better to optimize for the general case rather than the uncommon one.

davecgh commented on callbacks.go:
```diff
@@ -0,0 +1,605 @@
+
+// pushTxMsg sends a tx message for the provided transaction hash to the
+// connected peer.  An error is returned if the transaction hash is not known.
+var pushTxMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Any reason this is a var instead of just:

```Go
func pushTxMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```

tuxcanfly replied:
I thought it might be useful for testing, will recheck that when writing tests.

davecgh commented on callbacks.go:
```diff
@@ -0,0 +1,605 @@
+
+// pushBlockMsg sends a block message for the provided block hash to the
+// connected peer.  An error is returned if the block hash is not known.
+var pushBlockMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Same here.  No need for variable.

davecgh commented on callbacks.go:
```diff
@@ -0,0 +1,605 @@
+// the connected peer.  Since a merkle block requires the peer to have a filter
+// loaded, this call will simply be ignored if there is no filter loaded.  An
+// error is returned if the block hash is not known.
+var pushMerkleBlockMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Same here.

davecgh commented:
Also, before this is ready for merge it will need the standard fare of package documentation.

- [ ] doc.go
- [ ] README.md
- [ ] example_test.go

tuxcanfly commented:
Thanks @davecgh will add those.

tuxcanfly commented:
Refactored to remove fields owned by blockmanager i.e `requestedTxns`, `requestedBlocks` from `peer.Peer`. Updated blockmanager to maintain a map of requested objects to peers. Now testing to make sure this works as expected.

As for pending tasks, added a first version of `README.md`. `doc.go`. Need to add `example.go` and tests for package `peer`.

tuxcanfly commented:
Added example to showcase usage of package `peer`. Also fixed the bugs around `btcctl` peer commands.

Working on testing and writing `peer` tests.

tuxcanfly commented:
Added both inbound and outbound peer examples. Working on ironing out a couple of minor issues and adding tests.

tuxcanfly commented:
Added tests for package peer, with coverage at ~47% now. Working on clearing up and improving the coverage.

tuxcanfly commented:
Debugging an crash in `handleInvMsg` which occurs due to nil IV.

tuxcanfly commented:
Pushed a fix for the inv message crash issue.

davecgh commented on blockmanager.go:
```diff
@@ -425,16 +429,20 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, p *peer) {
-	for k := range p.requestedTxns {
-		delete(b.requestedTxns, k)
+	for k, txPeer := range b.requestedTxns {
+		if txPeer == p { // peer matches
```
I believe we have unique IDs for peers already.  We should probably use that instead of relying on pointer equality.  I realize that as long as the code is super careful they should match, but it wouldn't be very difficult to have two different pointers to the same peer which would cause a check like this fail.

davecgh commented on blockmanager.go:
```diff
@@ -425,16 +429,20 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, p *peer) {
-	for k := range p.requestedBlocks {
-		delete(b.requestedBlocks, k)
+	for k, blockPeer := range b.requestedBlocks {
+		if blockPeer == p { // peer matches
```
Same here regarding the ID.

davecgh commented on blockmanager.go:
```diff
@@ -425,16 +429,20 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, p *peer) {
 	// be fetched from elsewhere next time we get an inv.
-	for k := range p.requestedTxns {
-		delete(b.requestedTxns, k)
+	for k, txPeer := range b.requestedTxns {
```
I think the structure should categorize the requests by peer.  The intent for most of these is to be able to query information based on a peer and looping everything for every peer to see if the peer matches is quite inefficient.

Perhaps something like the following where int64 is the unique peer ID.
```Go
requestedTxns:   make(map[int64]map[wire.ShaHash]struct{})
```

Then adding a peer would be something like  `requestedTxns[p.ID()] = make(map[wire.ShaHash]struct{})` and looping it here would be a simple matter of: `for k := range b.requestedTxns[p.ID()] {`.

davecgh replied:
Also, in regards to this, there really needs to be an overall map in addition to the described one.  That way it's trivial to look up by requested hash or peer.  So something like:

```Go
requestedTxns:   make(map[wire.ShaHash]struct{}),
requestedPeerTxns:   make(map[int64]map[wire.ShaHash]struct{}),
```

davecgh replied:
An alternative idea here is to wrap the new `peer.Peer` in a btcd-specific peer that adds the peer-specific data it needs.  Something like:

```Go
type btcdPeer struct {
	*peer.Peer
	requestedTxns map[wire.ShaHash]struct{})
	...
}
```

However, I suspect the original suggestion is more efficient since it won't require the extra wrapping nor mapping of `peer.Peer` to btcdPeer that would have to be done in each of the callbacks.

davecgh commented on blockmanager.go:
```diff
@@ -480,7 +488,6 @@ func (b *blockManager) handleTxMsg(tmsg *txMsg) {
 	// instances of trying to fetch it, or we failed to insert and thus
 	// we'll retry next time we get an inv.
 	txHash := tmsg.tx.Sha()
-	delete(tmsg.peer.requestedTxns, *txHash)
```
This still need to be removed.  If the above advice is followed it would be:

```Go
if requestedTxns, ok := b.requestedTxns[p.ID()]; ok {
	delete(requestedTxns, *txHash)
}
```

davecgh replied:
Status on this piece?

tuxcanfly replied:
We're using `peerInfo.requestedTxns` and holding all peers in `blockManager.peers` so we can easily delete the tx from the peer requests queue.

davecgh commented on blockmanager.go:
```diff
@@ -534,7 +541,8 @@ func (b *blockManager) current() bool {
 	// If we didn't ask for this block then the peer is misbehaving.
 	blockSha := bmsg.block.Sha()
-	if _, ok := bmsg.peer.requestedBlocks[*blockSha]; !ok {
+
```
No need for the extra space here.

davecgh commented on blockmanager.go:
```diff
@@ -534,7 +541,8 @@ func (b *blockManager) current() bool {
 	blockSha := bmsg.block.Sha()
-	if _, ok := bmsg.peer.requestedBlocks[*blockSha]; !ok {
+
+	if p, exists := b.requestedBlocks[*blockSha]; !exists || p != bmsg.peer {
```
This would have to be updated when changed per above.  Something like:
```Go
if p.ID() == bmsg.peer.ID() || b.hasRequestedBlock(p) {
// ...
}
```
Where that `b.hasRequestedBlock` is a new convenience func used throughout as needed and would look something like:
```Go
func (b *blockManager) hasRequestedBlock(p *peer.Peer) bool {
	requestedBlocks, ok := b.requestedTxns[p.ID()]
	if !ok {
		return false
	}

	_, ok := requestedBlocks[p.ID()]
	return ok
}
```

davecgh commented on blockmanager.go:
```diff
@@ -572,10 +580,9 @@ func (b *blockManager) handleBlockMsg(bmsg *blockMsg) {
+	// Remove block from request map. Either chain will know about it and
 	// so we shouldn't have any more instances of trying to fetch it, or we
 	// will fail the insert and thus we'll retry next time we get an inv.
-	delete(bmsg.peer.requestedBlocks, *blockSha)
```
Same as above.

davecgh commented on blockmanager.go:
```diff
@@ -692,8 +699,14 @@ func (b *blockManager) handleBlockMsg(bmsg *blockMsg) {
+		}
 		if b.startHeader != nil &&
-			len(bmsg.peer.requestedBlocks) < minInFlightBlocks {
+			peerRequestedBlocks < minInFlightBlocks {
```
Same thing here.  When you change the structures to categorize, this can remain a simple len instead of having to do the counting dance above.

davecgh commented on blockmanager.go:
```diff
@@ -764,8 +777,7 @@ func (b *blockManager) fetchHeaderBlocks() {
 		if !haveInv {
-			b.requestedBlocks[*node.sha] = struct{}{}
-			b.syncPeer.requestedBlocks[*node.sha] = struct{}{}
+			b.requestedBlocks[*node.sha] = b.syncPeer
```
Naturally this will change too.

tuxcanfly replied:
Yeah, will refactor the structs to categorize and update all places where they're used.

davecgh commented on blockmanager.go:
```diff
@@ -1063,7 +1073,7 @@ func (b *blockManager) handleInvMsg(imsg *invMsg) {
 		}
 	}
-	imsg.peer.requestQueue = requestQueue
+	imsg.peer.SetRequestQueue(requestQueue)
```
Once the rest is changed, it'll likely affect this, so we can revisit then, but this doesn't feel right having something to muck with the internals of the peer.

davecgh commented on log.go:
```diff
@@ -11,6 +11,7 @@ import (
 
 	"github.com/btcsuite/btcd/addrmgr"
+	"github.com/btcsuite/btcd/peer"
 
```
Extra line.

davecgh commented on peer.go:
```diff
@@ -1,516 +1,518 @@
-
-const (
-	// maxProtocolVersion is the max protocol version the peer supports.
-	maxProtocolVersion = 70002
```
We should discuss this a bit.  I suspect we'll need both the max peer version in the peer package and a max server protocol version in btcd proper.  The reason is that with them being separate it is entirely possible the peer supports a higher version number than the server.

tuxcanfly replied:
OK, will keep both server and peer supported versions.

davecgh commented on peer.go:
```diff
@@ -1,516 +1,518 @@
+// handleVersionMsg returns a handler which is invoked when a peer
+// receives a version bitcoin message and is used to negotiate the
+// protocol version details as well as kick start the communications.
+func handleVersionMsg(s *server) func(p *peer.Peer, msg *wire.MsgVersion) {
```
Why return an additional closure on all of these?  The functions themselves have the same signature, so you can just define them on the server and then pass them directly.

```Go
func (s *server) handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) { ... }
...
p.AddVersionMsgListener("handleVersionMsg", s.handleVersionMsg)
```

tuxcanfly replied:
I thought it might be useful to allow composition to provide any other context that might be required later, e.g like in: https://gist.github.com/tsenart/5fc18c659814c078378d

Although I agree that it might be simpler to keep them on the `Server`, will check and refactor that.

davecgh commented on peer.go:
```diff
@@ -1,516 +1,518 @@
+		numAdded := 0
+		notFound := wire.NewMsgNotFound()
+
+		// We wait on the this wait channel periodically to prevent queueing
```
I realize this was already there, but might as well fix the comment while here.  `the this` -> `this`

davecgh commented on peer.go:
```diff
@@ -1,516 +1,518 @@
 // pushTxMsg sends a tx message for the provided transaction hash to the
 // connected peer.  An error is returned if the transaction hash is not known.
-func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+var pushTxMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Any reason this is a variable?

```Go
func pushTxMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
would allow the compiler to be more efficient.

davecgh commented on peer.go:
```diff
@@ -533,8 +535,8 @@ func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) er
 // connected peer.  An error is returned if the block hash is not known.
-func (p *peer) pushBlockMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
-	blk, err := p.server.db.FetchBlockBySha(sha)
+var pushBlockMsg = func(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Same

davecgh commented on peer.go:
```diff
@@ -553,7 +555,7 @@ func (p *peer) pushBlockMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{})
 	// an inv straight after.
 	var dc chan struct{}
-	sendInv := p.continueHash != nil && p.continueHash.IsEqual(sha)
+	sendInv := p.ContinueHash() != nil && p.ContinueHash().IsEqual(sha)
```
Seems like the continue hash is a feature of the handling code and thus shouldn't be defined on the peer.

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,28 @@
+ */
+
+// Package peer provides a common base for creating and managing bitcoin
+// network peers for full nodes, Simplified Payment Verification (SPV) nodes,
```
maybe `full nodes` -> `fully validating`?

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,28 @@
+// Package peer provides a common base for creating and managing bitcoin
+// network peers for full nodes, Simplified Payment Verification (SPV) nodes,
+// proxies etc. It handles all the common message handlers like initial message
+// version negotiation, handling and responding to pings. For other messages,
```
`negotiation, handling` -> `negotiation and handling`

davecgh commented on peer/example/inboundpeer/main.go:
```diff
@@ -0,0 +1,79 @@
+
+	addrMgr := addrmgr.New("test", lookupFunc)
+
+	logger, err := btclog.NewLoggerFromWriter(os.Stdout, btclog.DebugLvl)
```
It's generally better to use the following to ensure the log is properly flushed on program shutdown as well as provide a subsystem prefix:

```Go
backendLogger := btclog.NewDefaultBackendLogger()
defer backendLogger.Flush()
logger := btclog.NewSubsystemLogger(backendLogger, "PEER: ")
logger.SetLevel(btclog.DebugLvl)
peer.UseLogger(logger)
```

davecgh commented on peer/example/inboundpeer/main.go:
```diff
@@ -0,0 +1,79 @@
+		p.Shutdown()
+	})
+
+	<-done
```
This might be a little more consistent as `p.WaitForShutdown()` much like the btcd server does and have the peer handle the channel portion rather than accepting one.

davecgh commented on peer/log.go:
```diff
@@ -0,0 +1,247 @@
+	// is generated per 10 minutes, this allows blocks for about 9,512
+	// years.  However, if the field is interpreted as a timestamp, given
+	// the lock time is a uint32, the max is sometime around 2106.
+	lockTimeThreshold uint32 = 5e8 // Tue Nov 5 00:53:20 1985 UTC
```
This is exported by `txscript` as `txscript.LockTimeThreshold`.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+)
+
+// zeroHash is the zero value hash (all zeros).  It is defined as a convenience.
+var zeroHash wire.ShaHash
```
Should probably combine this with the `var` declaration block above.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	DisableListen    bool
+	Oniondial        func(string, string) (net.Conn, error)
+	Dial             func(string, string) (net.Conn, error)
+	Onread           func(int, *wire.Message, error)
```
`OnRead`

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	Oniondial        func(string, string) (net.Conn, error)
+	Dial             func(string, string) (net.Conn, error)
+	Onread           func(int, *wire.Message, error)
+	Onwrite          func(int, *wire.Message, error)
```
`OnWrite`

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+// example, .onion addresses will be dialed using the onion specific proxy if
+// one was specified, but will otherwise use the normal dial function (which
+// could itself use a proxy or not).
+func btcdDial(cfg *Config, network, address string) (net.Conn, error) {
```
This is only used internally and everything that calls it uses tcp for the network.  We should probably get rid of the effectively unused parameter.

davecgh replied:
In fact, perhaps we probably should think about not even have the peer doing the dialing at all.  It's much more flexible to simply accept a `net.Conn` interface and let the caller handle the connection aspects.  As long as the caller is notified when the peer disconnects, it can simply do the reconnect and create a new peer, which is effectively what it's doing anyways.

tuxcanfly replied:
OK will do that, we are already passing the `conn` for outbound peer.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+// It is used as a callback to get newest block details.
+type ShaFunc func() (sha *wire.ShaHash, height int64, err error)
+
+// Peer provides a bitcoin peer for handling bitcoin communications.  The
```
This comment needs to be updated to reflect reality.  Some of it will need to move to the block manager.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	id                 int32
+	inbound            bool
+	persistent         bool
+	KnownAddresses     map[string]struct{}
```
Any reason this is exported?

tuxcanfly replied:
We need to update this from addr message handler.

davecgh replied:
I think we'll need an update function that is protected by a mutex  for that.  Raw unprotected access to a map isn't concurrent safe.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	KnownAddresses     map[string]struct{}
+	knownInventory     *MruInventoryMap
+	knownInvMutex      sync.Mutex
+	retryCount         int64
```
If we don't do any of the reconnecting and such, this won't be needed.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+}
+
+// RequestQueue returns the peer request queue.
+func (p *Peer) RequestQueue() []*wire.InvVect {
```
I believe I commented elsewhere too, but allowing direct access to the request queue like this and in `SetRequestQueue` smells like a layer violation that is ripe for issues.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+
+// ContinueHash returns the continue hash.
+// It is safe for concurrent access.
+func (p *Peer) ContinueHash() *wire.ShaHash {
```
The continue hash is specific to the full block downloading mechanism and isn't directly related to the peer.  This is another piece of data that should be tracked by the block manager.

Given these various pieces of information, I suspect it might make sense to create another struct in btcd to house additional peer data and have a single `map[int64]*peerData` on the block manager which houses it all.

tuxcanfly replied:
Thanks, will do that. I think the above mentioned refactor for `requestedTxns` `requestedBlocks` could also be included in the new struct, `peerData`

davecgh replied:
Yep, that's what I was thinking as well.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+
+// RelayTxDisabled returns whether or not relaying of transactions is disabled.
+// It is safe for concurrent access.
+func (p *Peer) RelayTxDisabled() bool {
```
This is another thing that is not unique to the peer and should be in the aforementioned data struct in the block manager.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+
+// TxProcessed sends a message that a transaction has been processed by the
+// peer.
+func (p *Peer) TxProcessed() {
```
Another case where the data should be in the aforementioned block manager struct instead of on the peer.  This is very specific to the block manager's operation and not something a generic peer should have.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+}
+
+// BlockProcessed sends a message that a block has been processed by the peer.
+func (p *Peer) BlockProcessed() {
```
Same here.  Should be in the block manager struct instead of on the peer.  This is very specific to the block manager's operation and not something a generic peer should have.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	msg.AddrYou.Services = wire.SFNodeNetwork
+
+	// Advertise that we're a full node.
+	msg.Services = wire.SFNodeNetwork
```
This will need to be configurable as the peer is intended to support more than just full node use.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+func (p *Peer) PushGetBlocksMsg(locator blockchain.BlockLocator, stopHash *wire.ShaHash) error {
+	// Extract the begin hash from the block locator, if one was specified,
+	// to use for filtering duplicate getblocks requests.
+	// request.
```
Looks like a spurious comment line here.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+}
+
+// PushRejectMsg sends a reject message for the provided command, reject code,
+// and reject reason, and hash.  The hash will only be used when the command
```
I think it was already there, but might as well fix it while we're here.  There is an extra `and` before `reject reason`.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+
+// AddVersionMsgListener adds a listener which is invoked when a peer receives
+// a version bitcoin message.
+func (p *Peer) AddVersionMsgListener(key string, listener func(p *Peer, msg *wire.MsgVersion)) {
```
We either need to protect the callback registration and invocation by a mutex or clearly document that registration and invocation is not concurrent-safe.

tuxcanfly replied:
OK, will add a mutex.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+			p.StatsMtx.Unlock()
+
+			if !versionSent {
+				log.Infof("Received 'verack' from peer %v "+
```
I don't think this should be `Infof`.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+			}
+
+		case *wire.MsgNotFound:
+			// TODO(davec): Ignore this for now, but ultimately
```
Even though it's not used yet, we should expose a listener for this message.  It will be needed when we implement concurrent block downloading.  Also, other callers might need to react to it.

The comment can then be moved into the registration code in btcd proper.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+			}
+
+		case *wire.MsgReject:
+			// Nothing to do currently.  Logging of the rejected
```
Same here.  We should expose a listener for all supported messages even if btcd doesn't currently directly use them since this peer is meant for more than just btcd.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+			// message is handled already in readMessage.
+
+		default:
+			log.Debugf("Received unhandled message of type %v: Fix Me",
```
Please remove the `: Fix Me`.  That was missed and should no longer be there.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+	p.donePeers <- p
+
+	// Only tell block manager we are gone if we ever told it we existed.
+	// TODO
```
Reminder

davecgh commented:
I'll finish the review tomorrow since it's really late already.  This PR is definitely coming along nicely from what I've seen so far.  Nice work on it @tuxcanfly .

tuxcanfly commented:
Thanks @davecgh Let me get those items addressed and prep up for further review.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+// Start begins processing input and output messages.  It also sends the initial
+// version message for outbound connections to start the negotiation process.
+func (p *Peer) Start() error {
+
```
Extra space

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+// Disconnect disconnects the peer by closing the connection.  It also sets
+// a flag so the impending shutdown can be detected.
+func (p *Peer) Disconnect() {
+	// did we win the race?
```
Please nuke this unhelpful comment.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+func NewInboundPeer(NewestSha ShaFunc, nonce uint64, donePeers chan *Peer,
+	btcnet wire.BitcoinNet, addrManager *addrmgr.AddrManager, cfg *Config, conn net.Conn) *Peer {
+	p := newPeerBase(NewestSha, nonce, donePeers,
+		btcnet, addrManager, cfg, true)
```
The wrapping here isn't consistent.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+func NewOutboundPeer(NewestSha ShaFunc, nonce uint64, donePeers chan *Peer,
+	btcnet wire.BitcoinNet, addrManager *addrmgr.AddrManager, cfg *Config, addr string, persistent bool, retryCount int64) *Peer {
+	p := newPeerBase(NewestSha, nonce, donePeers,
+		btcnet, addrManager, cfg, false)
```
Same thing here.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,1942 @@
+// newPeerBase returns a new base bitcoin peer for the provided server and
+// inbound flag.  This is used by the NewInboundPeer and NewOutboundPeer
+// functions to perform base setup needed by both types of peers.
+func newPeerBase(NewestSha ShaFunc, nonce uint64, donePeers chan *Peer,
```
NewestSha shouldn't be capitalized as that goes against effective Go function variable style guidelines.  The same is true for the other cases.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,290 @@
+// Copyright (c) 2013-2014 The btcsuite developers
```
2015

davecgh commented on rpcserver.go:
```diff
@@ -76,6 +76,9 @@ const (
 	// in the memory pool.
 	gbtRegenerateSeconds = 60
+
+	// maxProtocolVersion is the max protocol version the peer supports.
```
`the server supports`

davecgh commented on server.go:
```diff
@@ -302,14 +328,35 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+					// Signal the block manager this peer is a new sync candidate.
+					s.blockManager.NewPeer(p)
+				}
+				p.AddVersionMsgListener("handleVersionMsg", handleVersionMsg)
```
Isn't this (and the code before it) already done by the call to `s.registerListeners(p)` below?  If not, they're using the same key so it would be overwritten anyways.

tuxcanfly replied:
Oops, should have been removed after adding `s.registerListeners(p)`

davecgh commented on server.go:
```diff
@@ -446,32 +493,31 @@ func (s *server) handleQuery(querymsg interface{}, state *peerState) {
 			// and we don't really care if they are raced to get the new
 			// version.
-			p.StatsMtx.Lock()
+			// TODO(tuxcanfly): check and restore stats mtx lock
```
Reminder!

davecgh commented:
Alright, I'm done with this round of review.  If you don't mind, please squash everything up to commit c734521dec2ccc892354b61cc93665a6b9f2aee5, so it's easier to review the incremental changes that are being made to address the review items.  Thanks!

I'll give it another round of review once everything is addressed.  Also, I will spend some time to do some thorough testing once the PR is nearing completion.

davecgh commented:
For reference, I will be pushing a rather large infrastructure changing PR somewhere in the time frame of a week.  There will likely be several conflicts with this PR as it is also quite far reaching.

I'd like to get this to a point it's ready for merge before the other PR drops so I can deal with the merge conflicts at that time.  Otherwise, you'd have to deal with them in this PR which, I believe will be more difficult than the other way around due to the nature of the changes.

tuxcanfly commented:
Thanks for the heads up @davecgh  Let me work on getting it ready to merge.

tuxcanfly commented:
@davecgh Good for another round, I've addressed almost all of the items addressed earlier. Still working on tests.

davecgh commented on blockmanager.go:
```diff
@@ -221,6 +238,17 @@ func (b *blockManager) resetHeaderState(newestHash *wire.ShaHash, newestHeight i
+// peer is disabled.
+// It is safe for concurrent access.
+func (b *blockManager) RelayTxDisabled(p *peer.Peer) bool {
+	pd := b.peers[p.Id()]
```
This should be checking if the peer entry exists and warn if it doesn't.  Currently, it will just panic and take down the entire server when given a peer that is no longer in the map.  Also, I think the map access should be protected for it to be concurrent safe as well.


For the warning part, something like:

```Go
data, ok := b.peers[p.ID()]
if !ok {
	log.Warnf(....)
}
...
```

tuxcanfly replied:
Thanks, yeah calling this before initializing the peer in `handleNewPeerMsg` will panic, let me add the check. Also agreed about the mutex map, let me add that.

davecgh commented:
> $ golint
> peer.go:309:16: func Id should be ID

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,28 @@
+// Creating a peer requires a method to retrieve the newest block sha, a nonce,
+// a channel to signal peer completion, an address manager and customizable
+// config. Once initialized, a peer can be started using the Start method.
+
```
There can't be a space here or the documentation will not show up.  I'd suggest using block comments as well for consistency.

```Go
/*
Package peer ...

Creating a peer ....
*/
package peer
```

davecgh commented on peer/log.go:
```diff
@@ -0,0 +1,241 @@
+// Copyright (c) 2013-2014 The btcsuite developers
```
2015

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,164 @@
+func handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) {
+	var zeroHash wire.ShaHash
+	locator := blockchain.BlockLocator([]*wire.ShaHash{&zeroHash})
+	p.PushGetBlocksMsg(locator, &zeroHash)
```
I realize this is just test code, but I'd rather the test code is realistic.  Pushing a get blocks message in response to a version message goes against the protocol.

tuxcanfly replied:
OK, let me update it to follow the protocol.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+	// base done the number of retries that have been done.
+	maxConnectionRetryInterval = time.Minute * 5
+
+	defaultLogFilename = "btcd.log"
```
This looked like an accidental paste since it's not used in the peer package (nor should it be).

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+	// and is used to assign an id to a peer.
+	nodeCount int32
+
+	// zeroHash is the zero value hash (all zeros).  It is defined as a convenience.
```
Comment exceeds 80.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+	zeroHash wire.ShaHash
+)
+
+// Config is the struct to hold configuration options useful to Peer
```
Since this is an exported type that callers need to interact with, I think each field should be documented.  Perhaps the [documentation](https://github.com/btcsuite/btcrpcclient/blob/master/infrastructure.go#L1016-L1079) on the `ConnConfig` type in [btcrpcclient](https://github.com/btcsuite/btcrpcclient/) would be a useful guide.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+	SimNet           bool
+	RegressionTest   bool
+	DisableListen    bool
+	Onread           func(int, *wire.Message, error)
```
`OnRead` better matches casing rules.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+	RegressionTest   bool
+	DisableListen    bool
+	Onread           func(int, *wire.Message, error)
+	Onwrite          func(int, *wire.Message, error)
```
`OnWrite` better matches casing rules.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2098 @@
+
+const (
+	// maxProtocolVersion is the max protocol version the peer supports.
+	maxProtocolVersion = 70002
```
I wonder if this should be exported now since callers might want to know the highest version the peer package supports.

davecgh replied:
Thoughts on this?

tuxcanfly replied:
OK, exported.

davecgh commented on blockmanager.go:
```diff
@@ -186,23 +190,36 @@ type blockManager struct {
+
+	continueHash   *wire.ShaHash
+	disableRelayTx bool
+	relayMtx       sync.Mutex
```
Super minor nitpick, but typically a mutex is defined before the thing(s) it protects.

davecgh commented on blockmanager.go:
```diff
@@ -423,19 +457,9 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, p *peer) {
-	// fetched from elsewhere next time we get an inv.
-	// TODO(oga) we could possibly here check which peers have these blocks
-	// and request them now to speed things up a little.
-	for k := range p.requestedBlocks {
```
This, and the previous for requstedTxns, is still going to be needed.  It will just be done against the peer data now.

davecgh commented on blockmanager.go:
```diff
@@ -480,8 +510,9 @@ func (b *blockManager) handleTxMsg(tmsg *txMsg) {
 	txHash := tmsg.tx.Sha()
-	delete(tmsg.peer.requestedTxns, *txHash)
-	delete(b.requestedTxns, *txHash)
+	for _, p := range b.peers {
```
This should only be deleting from the specific peer that requested it, not all of them.

davecgh commented on blockmanager.go:
```diff
@@ -524,7 +555,7 @@ func (b *blockManager) current() bool {
 	// parse an orphan, which would allow us to update the height of peers
 	// from what it was at initial handshake.
-	if err != nil || height < int64(b.syncPeer.lastBlock) {
+	if err != nil || height < int64(b.syncPeer.LastBlock()) {
```
`LastAnnouncedBlock`

tuxcanfly replied:
`LastAnnouncedBlock` returns a hash, not the height which is required here.

davecgh replied:
Ah right.  Well this check needs to compare against the latest announced height to be correct.  Unless I've missed something with your new code, the `LastBlock()` is not updated after the initial connection.  This check is for selecting a new sync peer and therefore needs to be comparing against the latest known height as ascertained from their inventory announcements.

davecgh commented on blockmanager.go:
```diff
@@ -1041,9 +1075,8 @@ func (b *blockManager) handleInvMsg(imsg *invMsg) {
-			if _, exists := b.requestedBlocks[iv.Hash]; !exists {
-				b.requestedBlocks[iv.Hash] = struct{}{}
-				imsg.peer.requestedBlocks[iv.Hash] = struct{}{}
+			if _, exists := b.peers[imsg.peer.ID()].requestedBlocks[iv.Hash]; !exists {
```
Rather than looking up `b.peers[imsg.peer.ID()]` multiple times throughout the function, why not just get a pointer to the peer data once and use the local throughout the function?

```Go
peerState := b.peers[imsg.peer.ID()]
...
case wire.InvTypeBlock:
	_, exists := peerState.requestedblocks[iv.Hash]
	if !exists {
		...
```

davecgh commented on blockmanager.go:
```diff
@@ -1063,7 +1095,7 @@ func (b *blockManager) handleInvMsg(imsg *invMsg) {
 		}
 	}
-	imsg.peer.requestQueue = requestQueue
+	b.peers[imsg.peer.ID()].requestQueue = requestQueue
```
If you implement my previous suggestion this just becomes:

```Go
peerState.requestQueue = requestQueue
```

So you avoid another map lookup.

davecgh commented on blockmanager.go:
```diff
@@ -1084,14 +1116,15 @@ out:
 			switch msg := m.(type) {
 			case *newPeerMsg:
 				b.handleNewPeerMsg(candidatePeers, msg.peer)
+				b.PeerProcessed()
```
This isn't used anywhere else and is only a single line to signal the channel.  I'd just inline it.

davecgh commented on blockmanager.go:
```diff
@@ -1084,14 +1116,15 @@ out:
 			case *txMsg:
 				b.handleTxMsg(msg)
-				msg.peer.txProcessed <- struct{}{}
+				b.TxProcessed()
```
Same here.

davecgh replied:
This needs to be a channel in the peer data instead of the block manager as it's unique to the peer that is processing it.  As you've changed it currently, you've made it so getting a tx or block from *any* peer will cause the code that is blocking to erroneously continue.

davecgh commented on blockmanager.go:
```diff
@@ -1084,14 +1116,15 @@ out:
 			case *blockMsg:
 				b.handleBlockMsg(msg)
-				msg.peer.blockProcessed <- struct{}{}
+				b.BlockProcessed()
```
Same here.

davecgh commented on blockmanager.go:
```diff
@@ -1168,6 +1201,39 @@ out:
 
+// TxProcessed sends a message that a transaction has been processed by the
+// block manager.
+func (b *blockManager) TxProcessed() {
```
As mentioned above, I don't think this function is necessary.

davecgh commented on blockmanager.go:
```diff
@@ -1168,6 +1201,39 @@ out:
+
+// BlockProcessed sends a message that a block has been processed by the block
+// manager.
+func (b *blockManager) BlockProcessed() {
```
Nor this one.

davecgh commented on blockmanager.go:
```diff
@@ -1168,6 +1201,39 @@ out:
+
+// PeerProcessed sends a message that a peer has been processed by the block
+// manager.
+func (b *blockManager) PeerProcessed() {
```
Nor this one.

davecgh commented on blockmanager.go:
```diff
@@ -1168,6 +1201,39 @@ out:
+
+// WaitForTx waits until a transaction has been processed by the block manager.
+func (b *blockManager) WaitForTx() {
+	<-b.txProcessed
```
As mentioned above, this is incorrect.  You're signalling when *any* peer receives a transaction which is not the intention.  The intention is to wait until a specific peer fulfills the request.  For that, the channel needs to be in the peer data.

davecgh commented on peer.go:
```diff
@@ -1,516 +1,18 @@
 // pushTxMsg sends a tx message for the provided transaction hash to the
 // connected peer.  An error is returned if the transaction hash is not known.
-func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+func pushTxMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
This should be moved onto the server type in `server.go` imo.  It's only used by its `handleGetDataMsg` and seems to more logically fit there.  Plus, by moving this along with `pushBlockMsg` and `pushMerkleBlockMsg`, the entire `peer.go` file can be removed.

```diff
diff --git a/server.go b/server.go
index 4072f8d..03adecb 100644
--- a/server.go
+++ b/server.go
@@ -288,6 +288,33 @@ func (s *server) handleHeadersMsg(p *peer.Peer, msg *wire.MsgHeaders) {
        s.blockManager.QueueHeaders(msg, p)
 }

+// pushTxMsg sends a tx message for the provided transaction hash to the
+// connected peer.  An error is returned if the transaction hash is not known.
+func (s *server) pushTxMsg(p *peer.Peer, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+       ...
+}
+
 // handleGetData is invoked when a peer receives a getdata bitcoin message
 // and is used to deliver block and transaction information.
 func (s *server) handleGetDataMsg(p *peer.Peer, msg *wire.MsgGetData) {
@@ -313,7 +340,7 @@ func (s *server) handleGetDataMsg(p *peer.Peer, msg *wire.MsgGetData) {
```

davecgh commented on peer.go:
```diff
@@ -533,8 +35,8 @@ func (p *peer) pushTxMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) er
 // connected peer.  An error is returned if the block hash is not known.
-func (p *peer) pushBlockMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
-	blk, err := p.server.db.FetchBlockBySha(sha)
+func pushBlockMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Same deal here as with `pushTxMsg` regarding moving to `server.go` and defining it before `handleGetData`.

davecgh commented on peer.go:
```diff
@@ -583,16 +86,16 @@ func (p *peer) pushBlockMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{})
 // loaded, this call will simply be ignored if there is no filter loaded.  An
 // error is returned if the block hash is not known.
-func (p *peer) pushMerkleBlockMsg(sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
+func pushMerkleBlockMsg(p *peer.Peer, s *server, sha *wire.ShaHash, doneChan, waitChan chan struct{}) error {
```
Same deal here as with `pushTxMsg` regarding moving to `server.go` and defining it before `handleGetData`.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+}
+
+// Stats is the collection of stats related to a peer
+type Stats struct {
```
This structure is not used outside of the peer package and therefore should not be exported.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+
+// Stats is the collection of stats related to a peer
+type Stats struct {
+	StatsMtx           sync.Mutex // protects all statistics below here.
```
This shouldn't be exported either.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+
+	Stats
+
+	NewestSha   ShaFunc
```
Shouldn't be exported.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+	knownAddrMutex     sync.Mutex
+	knownInventory     *MruInventoryMap
+	knownInvMutex      sync.Mutex
+	prevGetBlocksBegin *wire.ShaHash // owned by blockmanager
```
These aren't owned by the block manager.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+
+// AddKnownAddress adds the passed address to the cache of known addresses
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownAddress(addr string) {
```
I think the addition of the known addresses should be handled by the peer itself before invoking the callback since the filtering it enables is something that should happen for all peers, regardless of the use case.

It would obviously need to be changed slightly from what is currently in the server handler.

davecgh replied:
Reminder for this.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+}
+
+// RetryCount returns the number of times to retry peer connection.
+func (p *Peer) RetryCount() int64 {
```
As mentioned in IRC, the connection retry code should be the responsibility of the caller, not the peer.  That means this `RetryCount` and the associated field can go away and move into the `peerData` struct on the server.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+	//      actually supports
+	//    - Set the remote netaddress services to the what was advertised by
+	//      by the remote peer in its version message
+	msg.AddrYou.Services = p.Services()
```
This is incorrect.  The remote peer services is not the same as the local peer.  It should be set to `SFNodeNetwork` as it was before.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2114 @@
+	//      by the remote peer in its version message
+	msg.AddrYou.Services = p.Services()
+
+	// Advertise that we're a full node.
```
Comment is wrong.

davecgh commented on server.go:
```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
 	s.modifyRebroadcastInv <- broadcastInventoryDel(iv)
 }
 
+// handleVersionMsg returns a handler which is invoked when a peer
```
Comment is incorrect.

davecgh commented on server.go:
```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+	pd.disableRelayTx = msg.DisableRelayTx
+}
+
+// handleMemPoolMsg returns a handler which is invoked when a peer receives a mempool bitcoin
```
This one too.

davecgh commented on server.go:
```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+	// is received.
+	pd := s.blockManager.peers[p.ID()]
+	pd.relayMtx.Lock()
+	defer pd.relayMtx.Unlock()
```
I'd not defer the unlock here since there is only one exist path and it's much cheaper to avoid the defer.

davecgh commented on server.go:
```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+		var err error
+		switch iv.Type {
+		case wire.InvTypeTx:
+			err = pushTxMsg(p, s, &iv.Hash, c, waitChan)
```
With the aforementioned changes, this would become:

```Go
s.pushTxMsg(p, &iv.Hash, c, waitChan)
```

The same for the `pushBlockMsg` and `pushMerkleBlockMsg` below.

davecgh commented on server.go:
```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+			// is not a reference into the inventory slice which
+			// would prevent the entire slice from being eligible
+			// for GC as soon as it's sent.
+			s.blockManager.peers[p.ID()].continueHash = &invMsg.InvList[invListLen-1].Hash
```
What happened to the hash copy?  As the comment states, it is needed to avoid delaying GC.

davecgh commented:
Also, please don't forget to add an `example_test.go` which provides (machine tested) example usage.

tuxcanfly commented:
Improved test coverage, added `example_test`, refactored address manager and retries to be handled outside peer.

dajohi commented:
connectNodeMsg needs to change; it blocks at btcdDial.

If you dial a non-responsive peer, you'll block other outbound attempts until the OS's system-wide timeout has expired.  Connection attempts should not be blocking.

dajohi commented:
seeing these too:
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0
17:12:29 2015-08-25 [WRN] BMGR: Missing peer data for peer id: 0

tuxcanfly commented:
OK, I think it was due to improper peer initialization due to the recent async connection commit, let me fix that.

davecgh commented on blockmanager.go:
```diff
@@ -221,6 +241,22 @@ func (b *blockManager) resetHeaderState(newestHash *wire.ShaHash, newestHeight i
+// RelayTxDisabled returns whether or not relaying of transactions for the given
+// peer is disabled.
+// It is safe for concurrent access.
+func (b *blockManager) RelayTxDisabled(p *peer.Peer) bool {
```
As mentioned in IRC, this will be changing to take a `*peerInfo` directly and the `peerInfo` struct will have a pointer to the peer.  I won't comment on all of the other cases, but just putting this here for reference.

davecgh commented on blockmanager.go:
```diff
@@ -203,6 +206,36 @@ type blockManager struct {
+	b.peersMtx.RLock()
+	pInfo, ok := b.peers[p.ID()]
+	if !ok {
+		return nil, errors.New("missing peer info")
```
Don't forgot to `b.peerMtx.RUnlock()` before returning.

tuxcanfly replied:
Thanks, added a `defer` for that.

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,18 @@
+/*
+
+Package peer provides a common base for creating and managing bitcoin network
+peers for fully validating, Simplified Payment Verification (SPV) nodes,
```
fully validating nodes

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,18 @@
+
+Package peer provides a common base for creating and managing bitcoin network
+peers for fully validating, Simplified Payment Verification (SPV) nodes,
+proxies etc. It handles all the common message handlers like initial message
```
proxies, etc

davecgh commented:
Need to update the copyright years for `peer/mruinvmap.go` to `2013-2015`.

davecgh commented on peer/example_test.go:
```diff
@@ -0,0 +1,86 @@
+	return nil, errors.New("not implemented")
+}
+
+// newestSha returns the latest known block to this peer
```
Need to end with a period for consistency.

davecgh commented on peer/example_test.go:
```diff
@@ -0,0 +1,86 @@
+	hash := "14a0810ac680a3eb3f82edc878cea25ec41d6b790744e5daeef"
+	height = 234439
+	sha, err = wire.NewShaHashFromStr(hash)
+	return
```
This is a good example of why we don't use named returns.  In this case, a height of 234439 would be returned instead of 0 if an error were to occur.

Also, we've been trying to work towards removing variables named `sha` from the code base since the hash might not always be a sha if the hashing function changes.

I would suggest the following:

```Go
// newestSha returns the latest known block to this peer.  It returns a
// hard-coded hash and height since this is only used for example code.
func newestSha() (*wire.ShaHash, int32, error) {
	hashStr := "14a0810ac680a3eb3f82edc878cea25ec41d6b790744e5daeef"
	hash, err := wire.NewShaHashFromStr(hashStr)
	if err != nil {
		return nil, 0, err
	}
	return hash, 234439, nil
}
```

tuxcanfly replied:
Thanks, will update that.

davecgh commented on peer/example_test.go:
```diff
@@ -0,0 +1,89 @@
+			return
+		}
+	}()
+	<-listening
```
This will block forever if there is an error establishing the listener in the goroutine (which could easily happen if btcd it running already listening on 8333, for example).

tuxcanfly replied:
Thanks, updated to use an `err` `chan` and handle the `err`

davecgh commented on peer/example_test.go:
```diff
@@ -0,0 +1,93 @@
+		if err != nil {
+			log.Fatalf("Listen: error %v\n", err)
+		}
+		p1 := peer.NewInboundPeer(newestSha, 0, wire.MainNet, wire.SFNodeNetwork, addrMgr, peerCfg, c1)
```
Given this is example code, I'd suggest being overly verbose in the comments to explain what's going on.  People looking at this code will be unfamiliar with the peer package and quite possibly Go itself.

davecgh replied:
Also seeing an example of this makes me think both `NewInboundPeer` and `NewOutboundPeer` probably should have a common type like a `PeerConfig` struct that they take.  The number of arguments is a bit unwieldy right now.  Typically once the arguments start to grow to this many, it's good practice to create a dedicated type for it.

tuxcanfly replied:
OK, will create a struct and pass the arguments using it.

davecgh replied:
I edited the message.  I noticed you already have the `Config` struct, I'd make use of that.  Perhaps the nonce should remain separate though as it's likely the caller will reuse the config struct and the nonce needs to be unique (random) per peer.

tuxcanfly replied:
OK, noted.

davecgh commented:
The test coverage is still pretty low on this:

$ go test -coverprofile=cov.out
PASS
coverage: 56.4% of statements

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+	// SOCKS5 proxy (eg. 127.0.0.1:9050) to use for connections
+	Proxy string
+
+	// Whether to use the imulation test network
```
simulation

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// Config is the struct to hold configuration options useful to Peer
+type Config struct {
+	// SOCKS5 proxy (eg. 127.0.0.1:9050) to use for connections
```
Period for consistency.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+	// Whether to use the imulation test network
+	SimNet bool
+
+	// Whether to use the regression test network
```
Period for consistency.

tuxcanfly commented:
OK, will work on improving the test coverage.

davecgh commented on peer/example_test.go:
```diff
@@ -0,0 +1,93 @@
+		fmt.Printf("Dial: error %v\n", err)
+		return
+	}
+	p2 := peer.NewOutboundPeer(newestSha, 1, wire.MainNet, wire.SFNodeNetwork, addrMgr, peerCfg, c2, "127.0.0.1:8333", true)
```
The nonce is supposed to be random per the spec.  The example should show that as well.

tuxcanfly replied:
Thanks, will update that.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,384 @@
+)
+
+// Mocks a network connection. Implements the net.Conn interface
+type Conn struct {
```
Any reason this (and `Pipe`) are exported?  They are only used within the test package.

davecgh replied:
I see you unexported `Pipe` but not this type and its fields.  Was that intentional?

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,18 @@
+// license that can be found in the LICENSE file.
+
+/*
+
```
Extra blank line here.

davecgh commented:
Please add an entry for this new package to the [Developer resources](https://github.com/btcsuite/btcd/blob/master/docs/README.md#DeveloperResources) section of [docs/README.md](https://github.com/btcsuite/btcd/tree/master/docs/README.md).

davecgh commented on peer/mruinvmap.go:
```diff
@@ -1,12 +1,13 @@
-// Copyright (c) 2013-2014 The btcsuite developers
+// Copyright (c) 2015 The btcsuite developers
```
Should be `2013-2015` since the file was moved and the code was updated.

davecgh commented on peer/mruinvmap_test.go:
```diff
@@ -1,8 +1,8 @@
-// Copyright (c) 2013-2014 The btcsuite developers
+// Copyright (c) 2015 The btcsuite developers
```
`2013-2015` here as well.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,384 @@
+}
+
+func handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) {
+	addrMgr.AddAddress(p.NA(), p.NA())
```
The intent of adding an address to the address manager is to add the remote address and the source address in order to help establish which ones are capable of communicating.  Since this is a test it's probably not a big deal to use the same one, but it definitely looks strange to see it here like this.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,384 @@
+}
+
+func handleMemPoolMsg(p *peer.Peer, msg *wire.MsgMemPool) {
+	invMsg := wire.NewMsgInvSizeHint(0)
```
No point in using the `SizeHint` variant with 0.  It will be 0 by default with just `NewMsgInv`.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,384 @@
+
+// testPeer tests the given peer's flags and stats
+func testPeer(t *testing.T, p *peer.Peer, inbound bool) {
+	var wantID, wantLastBlock, wantStartingHeight int32
```
You could save quite a bit of space here just assigning them the inbound values and changing the specific ones that are different for !inbound.  For example:

```Go
wantID := int32(1)
wantAddr := "127.0.0.1:8333"
...
wantInbound := true
...
if !inbound {
	wantID = 2
	...
	wantInbound = false
	...
```

davecgh commented on server.go:
```diff
@@ -160,6 +185,719 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+			// would prevent the entire slice from being eligible
+			// for GC as soon as it's sent.
+			continueHash := invMsg.InvList[invListLen-1].Hash
+			_, ok := s.blockManager.peers[p.ID()]
```
No need to do the map lookup a second time here:
```Go
peer, ok := s.s.blockManager.peers[p.ID()]
if !ok {
...
}
peer.continueHash = &continueHash
```

davecgh commented on server.go:
```diff
@@ -160,6 +185,719 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+// message is received.
+func (s *server) handleFilterAddMsg(p *peer.Peer, msg *wire.MsgFilterAdd) {
+	if !p.Filter().IsLoaded() {
+		peerLog.Debugf("%s sent a Filter()add request with no Filter() "+
```
Looks like a copy and replace error here with `Filter()`.

davecgh commented on server.go:
```diff
@@ -160,6 +185,719 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+// received.
+func (s *server) handleFilterClearMsg(p *peer.Peer, msg *wire.MsgFilterClear) {
+	if !p.Filter().IsLoaded() {
+		peerLog.Debugf("%s sent a Filter()clear request with no "+
```
Same here.

davecgh commented on server.go:
```diff
@@ -320,15 +1053,69 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+				if retryDuration > maxConnectionRetryInterval {
+					retryDuration = maxConnectionRetryInterval
+				}
+				state.retries[p] = retryDuration
```
This is going to race.  It's accessing state.retries unprotected form a separate goroutine.

davecgh commented on server.go:
```diff
@@ -320,15 +1053,69 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+				p.Config(), conn, p.Addr(), true)
+			s.registerListeners(p)
+
+			state.outboundGroups[addrmgr.GroupKey(p.NA())]++
```
This is a race too.  Pretty much everything in here is.

davecgh commented on server.go:
```diff
@@ -346,13 +1133,14 @@ func (s *server) handleRelayInvMsg(state *peerState, msg relayMsg) {
 			// Don't relay the transaction to the peer when it has
 			// transaction relaying disabled.
-			if p.RelayTxDisabled() {
+
```
Errant space.

davecgh commented on server.go:
```diff
@@ -302,14 +1037,12 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
-				e = newOutboundPeer(s, p.addr, true, p.retryCount+1)
-				list[e] = struct{}{}
+			if !p.Inbound() && p.Persistent() && atomic.LoadInt32(&s.shutdown) == 0 {
+				s.RetryPeer(p)
```
I don't see anything in the `RetryPeer` path updating the peer in the `state.peers` or `state.outboundPeers` lists, so the check above to see if the peer is one of the ones that should be retried due to being persistent will never be triggered give the `if e == p` check isn't going to match since it's a new peer pointer after the retry.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// stats is the collection of stats related to a peer
+type stats struct {
+	statsMtx           sync.Mutex // protects all statistics below here.
```
This should be a `RWMutex` imo.  The stats (particularly the peer ID) are accessed frequently from multiple goroutines, so they should be able to simultaneously read the information without blocking unless there is a write in progress.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// Addr returns the peer address.
+func (p *Peer) Addr() string {
+	p.flagsMtx.Lock()
```
The `addr` field doesn't change after the initial creation, thus there is no need to protect it with a mutex.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// Inbound returns whether the peer is inbound.
+func (p *Peer) Inbound() bool {
+	p.flagsMtx.Lock()
```
Same deal here.  The `inbound` field is assigned at creation time and never modified.  No need for mutex.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// Persistent returns whether the peer is persistent.
+func (p *Peer) Persistent() bool {
+	p.flagsMtx.Lock()
```
Same.  No need for mutex.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+
+// Config returns the configuration options of the peer.
+func (p *Peer) Config() *Config {
+	p.flagsMtx.Lock()
```
Same here.  `cfg` field never changed after initial creation.  No need for mutex.

davecgh commented on server.go:
```diff
@@ -446,32 +1234,30 @@ func (s *server) handleQuery(querymsg interface{}, state *peerState) {
-				Inbound:        p.inbound,
-				StartingHeight: p.startingHeight,
-				CurrentHeight:  p.lastBlock,
+				ID:             p.ID(),
```
Having to grab the mutex for each individual field here is not ideal.  I'd suggest a method on the peer like:

```Go
// StatsSnapshot returns a snapshot of the current peer statistics.
func (p *Peer) StatsSnapshot() *StatsSnap {
	p.statsMtx.RLock()
	defer p.statsMtx.RUnlock()

	// Copy the relevant fields into the snapshot...
	return &StatsSnap{
		ID: p.id,
		Addr: p.addr,
		// ...
	}
}
```

Then you could simply do here:

```Go
statsSnap := p.StatsSnapshot()
info := &btcjson.GetPeerInfoResult{
	ID: statsSnap.ID,
	// ...
	SyncNode: p == syncPeer,
}
```

davecgh replied:
In fact, now that peers are concurrent safe, this could be changed to simply return the list of peers and move all of the creation of the `btcjson.GetPeerInfoResult` into the rpc server where it belongs.

davecgh commented on server.go:
```diff
@@ -490,17 +1276,37 @@ func (s *server) handleQuery(querymsg interface{}, state *peerState) {
 		// TODO(oga) if too many, nuke a non-perm peer.
-		if s.handleAddPeerMsg(state,
-			newOutboundPeer(s, msg.addr, msg.permanent, 0)) {
+		peerCfg := &peer.Config{
```
This should be defined after the connection.  No reason to create it when the connection fails (which is quite often).  Once you've expanded `peer.Config` to include the connection, it would be more clear anyways.

davecgh commented on server.go:
```diff
@@ -673,7 +1491,28 @@ func (s *server) peerHandler() {
 	}
 	for _, addr := range permanentPeers {
-		s.handleAddPeerMsg(state, newOutboundPeer(s, addr, true, 0))
+		peerCfg := &peer.Config{
```
Same thing here regarding defining it after the connection is established.  Also, this code looks identical to the above code, so it should probably be refactored.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2043 @@
+// overall data flow is split into 3 goroutines and a separate block manager.
+// Inbound messages are read via the inHandler goroutine and generally
+// dispatched to their own handler.  For inbound data-related messages such as
+// blocks, transactions, and inventory, the data is passed on to the block
```
This comment is no longer accurate.  Need to scrub it and make it accurate.  For example, given it's a separate peer, information about the block manger is not true outside the scope of btcd..

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2027 @@
+}
+
+// LogError makes sure that we only log errors loudly on user peers.
+func (p *Peer) LogError(fmt string, args ...interface{}) {
```
There is no point in having this function once the persistent stuff was removed.  That said, it was intentionally doing different logging based upon whether it was a persistent peer, so you'll need to carefully examine why it was doing that and retrain (or otherwise) improve the behavior that made it necessary to begin with.

davecgh replied:
As an aside, if you run this branch currently, you can see why it was using debug logging.  It's now showing a bunch of peer errors that makes it look like something is wrong when, in reality, it is pretty typical for peers to come and go.  Those types of expected "errors" shouldn't be shown as errors which are more reserved for unexpected errors the user might need to report and or do something about.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2027 @@
+}
+
+// WaitForDisconnection waits until the peer is disconnected.
+func (p *Peer) WaitForDisconnection() {
```
This function is a duplicate of `WaitForShutdown`, it should be removed.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2027 @@
+}
+
+// queueHandler handles the queueing of outgoing data for the peer. This runs
+// as a muxer for various sources of input so we can ensure that blockmanager
```
Comment makes mention of block manager.  Needs to be updated appropriately.

tuxcanfly replied:
Done.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2013 @@
+type ShaFunc func() (sha *wire.ShaHash, height int32, err error)
+
+// Peer provides a basic bitcoin peer for handling bitcoin communications.  The
+// overall data flow is split into 3 goroutines.  Inbound messages are read via
```
More reference to block manager.  Also it is making a lot of references to internal function names.  This was fine before when it was not exported, but now this is a public comment, it needs to be changed accordingly.  That is what I meant previously by scrubbing it.  The information about the internal working should be split into a separate NOTE just above it since it is useful information for people digging into the peer package, but it doesn't belong in the exported comment.

Here is what I suggest:

```Go
// NOTE: The overall data flow of a peer is split into 3 goroutines.  Inbound
// messages are read via the inHandler goroutine and generally dispatched to
// their own handler.  For inbound data-related messages such as blocks,
// transactions, and inventory, the data is handled by the corresponding
// message handlers.  The data flow for outbound messages is split into 2
// goroutines, queueHandler and outHandler.  The first, queueHandler, is used
// as a way for external entities to queue messages, by way of the QueueMessage
// function, quickly regardless of whether the peer is currently sending or not.
// It acts as the traffic cop between the external world and the actual
// goroutine which writes to the network socket.

// Peer provides a basic concurrent safe bitcoin peer for handling bitcoin
// communications via the peer-to-peer protocol.  It provides full duplex
// reading and writing, automatic handling of the initial handshake process,
// querying of usage statistics and other information about the remote peer such
// as its address, user agent, and protocol version, output message queueing,
// inventory trickling, and the ability to dynamically register and unregister
// callbacks for handling bitcoin protocol messages.
//
// Outbound messages are typically queued via QueueMessage or QueueInventory. 
// QueueMessage is intended for all messages, including responses to data such
// as blocks and transactions.  QueueInventory, on the other hand, is only
// intended for relaying inventory as it employs a trickling mechanism to batch
// the inventory together.  However, some helper functions for pushing messages
// of specific types that typically require common special handling are
// provided as a convenience.
//
// The Add<Foo>Listener and Remove<Foo>Listener family of functions provide the
// ability for the caller to add and remove handlers, respectively, for each of
// the protocol messages the caller is interested in.
```

Notice the blank uncommented line in between the internal package note and the Peer comment to make sure the internal note does not appear in the doc for Peer.

tuxcanfly replied:
Thanks, updated.

davecgh commented on peer/peer_test.go:
```diff
@@ -0,0 +1,362 @@
+
+	if !inbound {
+		wantID = 2
+		wantAddr = "127.0.0.1:8333"
```
This and several of the other fields are already set to the same values above.  No need to override them to the exact same values.

davecgh commented on server.go:
```diff
@@ -320,15 +1055,65 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+		case <-time.After(retryDuration):
+			p, err := s.addPeer(p.Addr())
+			if err != nil {
+				srvrLog.Debugf("Failed to connect to %s: %v", p.Addr(), err)
```
This will crash if there is an error because p will be nil.

davecgh replied:
This will crash when there is an error because `p` will be `nil`.

davecgh commented on server.go:
```diff
@@ -320,15 +1055,65 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+					retryDuration = maxConnectionRetryInterval
+				}
+				state.retries[p] = retryDuration
+				s.RetryPeer(p)
```
Although it won't get this far due to the aforementioned crash, this would call `s.RetryPeer` with nil.

tuxcanfly replied:
Thanks, forgot to check this case when adding `addPeer`. Let me fix that.

tuxcanfly replied:
Fixed by not reusing variables.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2060 @@
+}
+
+// StatsSnapshot returns a snapshot of the current peer statistics.
+// It is used to return the result of a getpeerinfo peer message.
```
Given this is a package, we can't be sure how it's used by callers.  I'd remove this line.

tuxcanfly replied:
OK. removed.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2060 @@
+		Addr:      addr,
+		UserAgent: userAgent,
+		Services:  fmt.Sprintf("%08d", services),
+
```
Extra space.

tuxcanfly replied:
OK, fixed

tuxcanfly commented:
Tried a way around async peer connections and fixed concurrency issues around retries by refactoring. Working on moving out the remaining address manager bits and improving coverage.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2046 @@
+// UpdateLastBlockHeight updates the last known block for the peer. It is safe
+// for concurrent access.
+func (p *Peer) UpdateLastBlockHeight(newHeight int32) {
+	p.statsMtx.RLock()
```
This is changing the `p.lastBlock` field, so it can't be a read lock.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2046 @@
+
+	// Negotiate the protocol version.
+	p.protocolVersion = minUint32(p.protocolVersion, uint32(msg.ProtocolVersion))
+
```
No reason to have the blank line here.  It is all part of the negotiation.

davecgh commented on peer/peer.go:
```diff
@@ -0,0 +1,2046 @@
+		case *wire.MsgVerAck:
+			// No read lock is necessary because verAckReceived is not written
+			// to in any other goroutine
+			verAckReceived := p.verAckReceived
```
No need to make a copy.  Just access it below via `p.verAckReceived`.

davecgh commented on blockmanager.go:
```diff
@@ -534,7 +573,7 @@ func (b *blockManager) current() bool {
 	// If we didn't ask for this block then the peer is misbehaving.
 	blockSha := bmsg.block.Sha()
-	if _, ok := bmsg.peer.requestedBlocks[*blockSha]; !ok {
+	if _, exists := bmsg.peer.requestedBlocks[*blockSha]; !exists {
```
Please rename the field in the message if it's peer info.  This is confusing to read since it looks like a `*peer.Peer` instead of `*peerInfo`.

davecgh commented on blockmanager.go:
```diff
@@ -221,6 +248,28 @@ func (b *blockManager) resetHeaderState(newestHash *wire.ShaHash, newestHeight i
+
+	pInfo, ok := b.peers[p]
+	if !ok {
+		return nil, errors.New("missing peer info")
```
Please include some more context with this error.  Should it be hit, this error would be pretty useless for debugging.  Something like:

```Go
return nil, fmt.Errorf("missing peer info for %s", p)
```

davecgh commented on server.go:
```diff
@@ -639,6 +1469,23 @@ func (s *server) seedFromDNS() {
 	}
 }
 
+// addPeer adds a new outbound peer connecting to the given address.
```
Make sure to update this comment since it now longer makes the connection.

davecgh commented on server.go:
```diff
@@ -673,7 +1521,20 @@ func (s *server) peerHandler() {
+			if err != nil {
+				// Only log the error if we're not forcibly shutting down.
+				if atomic.LoadInt32(&s.shutdown) == 0 {
+					srvrLog.Errorf("can't accept connection: %v",
```
Not accepting a connection here, rather it's trying to establish one.

davecgh commented on server.go:
```diff
@@ -782,10 +1643,22 @@ out:
+				if err != nil {
+					// Only log the error if we're not forcibly shutting down.
+					if atomic.LoadInt32(&s.shutdown) == 0 {
+						srvrLog.Errorf("can't accept connection: %v",
```
Same here.  No accepting, establishing.

davecgh commented on server.go:
```diff
@@ -490,17 +1296,29 @@ func (s *server) handleQuery(querymsg interface{}, state *peerState) {
+			if err != nil {
+				// Only log the error if we're not forcibly shutting down.
+				if atomic.LoadInt32(&s.shutdown) == 0 {
+					srvrLog.Errorf("can't accept connection: %v",
```
Not accepting, establishing.

Also, I see the same code in 3 different places.  This seems ripe for a function.  Maybe something like:

```Go
// establishConn attempts to establish a connection with the passed peer.  It
// will typically need to be run as a goroutine to avoid blocking the caller.
func (s *server) establishConn(p *peer.Peer) {
	conn, err := btcdDial("tcp", p.Addr())
	// ....
	p.Connect(conn)
	return nil
}
```

Then, the code here would be:

```Go
p := s.addPeer(msg.addr)
go s.establishConn(p)
```

davecgh commented on server.go:
```diff
@@ -673,7 +1521,20 @@ func (s *server) peerHandler() {
 	for _, addr := range permanentPeers {
-		s.handleAddPeerMsg(state, newOutboundPeer(s, addr, true, 0))
+		p := s.addPeer(addr)
+		go func() {
```
And this would become:

```Go
go s.establishConn(p)
```

davecgh commented on server.go:
```diff
@@ -782,10 +1643,22 @@ out:
 
 			tries = 0
+			p := s.addPeer(addrStr)
+			go func() {
```
Same here:

```Go
go s.establishConn(p)
```

davecgh commented on server.go:
```diff
@@ -302,14 +1106,44 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+				// retrying peers.
+				go func() {
+					// Fetch last retry duration
+					retryDuration, ok := state.retries[p]
```
This will race.  Multiple goroutines accessing the state unprotected.

davecgh commented on server.go:
```diff
@@ -302,14 +1106,44 @@ func (s *server) handleDonePeerMsg(state *peerState, p *peer) {
+							if retryDuration > maxConnectionRetryInterval {
+								retryDuration = maxConnectionRetryInterval
+							}
+							state.retries[p] = retryDuration
```
Race.

tuxcanfly commented:
@davecgh Refactored and updated tests and docs. Please review when you're free.

tuxcanfly commented:
Rebased and squashed to include #501

davecgh commented:
Running `go test` is printing a bunch of debug messages.  That shouldn't be the case.

davecgh commented:
I was playing around with writing a toy application to use the `peer` package and noticed that `NewInboundPeer` takes a nonce.  I don't believe this should be the case.  An inbound peer is coming from somewhere and it will already have a nonce associated with it that said remote peer will deliver via the `version` message.

davecgh commented:
> // StatsSnap is a snapshot of peer stats at at point in time.

The word `at` is repeated.

davecgh commented:
`MaxProtocolVersion = 70002` needs to be bumped to 70011 since this incorporates the PR which bumps it for wire too.

jrick commented on blockmanager.go:
```diff
@@ -34,48 +37,100 @@ const (
+	p.relayMtx.Lock()
+	defer p.relayMtx.Unlock()
+
+	return p.disableRelayTx
```
I don't see anywhere in this PR where the disableRelayTx field is modified, so the mutex here looks unnecessary.  Is it here for a future change?

jrick replied:
I also don't see this function called anywhere.  Remove it as well as the disableRelayTx and relayMtx fields?

tuxcanfly replied:
This field is being modified from `server.go` in `handleVersionMsg` and `handleFilterLoadMsg`; the func is being called from `handleRelayInvMsg`

jrick replied:
Yeah GitHub is being stupid and didn't expand that file.  I'll continue reviewing this locally and not use the "files changed" view in this pr.

jrick commented on blockmanager.go:
```diff
@@ -34,48 +37,100 @@ const (
+// It is safe for concurrent access.
+func (p *peerInfo) addKnownAddress(addr string) {
+	p.knownAddrMutex.Lock()
+	defer p.knownAddrMutex.Unlock()
```
I would avoid the defer here since there's no complex branching that the defer helps to make readable and I don't trust the compiler to optimize this by avoiding the defer overhead.

tuxcanfly replied:
OK, will do.

jrick commented on blockmanager.go:
```diff
@@ -34,48 +37,100 @@ const (
+// It is safe for concurrent access.
+func (p *peerInfo) addressKnown(addr string) bool {
+	p.knownAddrMutex.Lock()
+	defer p.knownAddrMutex.Unlock()
```
see defer comment above.

tuxcanfly replied:
Thanks, updated both to remove `defer`.

jrick commented on blockmanager.go:
```diff
@@ -221,6 +278,19 @@ func (b *blockManager) resetHeaderState(newestHash *wire.ShaHash, newestHeight i
+
+	pInfo, ok := b.peers[p]
+	if !ok {
+		return nil, fmt.Errorf("Missing peer info for %s", p)
```
go error strings should be lowercase

tuxcanfly replied:
Thanks, updated.

davecgh commented:
The `// TODO: Relay alerts.` in `peer.go` can be nuked since that will not be the responsibility of the peer code.

davecgh commented on docs/README.md:
```diff
@@ -214,3 +214,5 @@ information.
 	  Provides a database interface for the Bitcoin block chain
     * [btcutil](https://github.com/btcsuite/btcutil) - Provides Bitcoin-specific
 	  convenience functions and types
+    * [peer](https://github.com/btcsuite/btcd/tree/master/peer) -
```
Would prefer this to come after the `wire` entry since it more logically fits there.

davecgh commented on rpcserver.go:
```diff
@@ -2277,7 +2281,38 @@ func handleGetNetworkHashPS(s *rpcServer, cmd interface{}, closeChan <-chan stru
+			BanScore:       0,
+			SyncNode:       p == syncPeer,
+		}
+		info.PingTime = float64(statsSnap.LastPingMicros)
```
Seems like this could be specified in the struct constructor.

davecgh commented:
As a point of reference, all previous comments up until this point have been taken care of.

tuxcanfly commented:
@davecgh Refactored to maintain peer state by reusing `peerState` maps. Please review when you're free. Thanks.

davecgh commented:
For reference, @tuxcanfly and I discussed this on IRC and he's going to refactor it to use a `serverPeer` instance closure in each new peer's config to remove the need for storing and retrieving the additional state in a map.  The new approach will be better since it can't run into a case where the entry is not in the map (which in turn could lead to other issues), is more efficient, and doesn't require additional locks.

tuxcanfly commented:
@davecgh Thanks, I've refactored to use `serverPeer` based listeners in peer config as discussed.

davecgh commented:
Nice.  I like this `serverPeer` refactor much better.  It is much cleaner in my opinion and results in a lot code changes.  I can't comment directly on the `server.go` file thanks to github not showing it for being too large, so I'll just combine the comments here:

---
```diff
 type serverPeer struct {
+       *server
```
This should be a field instead of embedding the entire server directly in the peer.

---
```Go
+func (sp *serverPeer) OnVersion(p *peer.Peer, msg *wire.MsgVersion) {
```
Since `sp.server.addrManager` is buried a couple of levels deep, I'd suggest creating an `addrManager := sp.server.addrManager` local and using that as I think It would make the code more readable.  The same holds for other handlers that are doing `sp.server.X` repeatedly.

---

```diff
-       // remote peer for outbound connections.  This is skipped when running
+       // remote peer for outbound connectionsp.server.  This is skipped when running
-       // discovered peers.
+       // discovered peersp.server.
-               // Outbound connections.
+               // Outbound connectionsp.server.
-                               // Get address that best matches.
+                               // Get address that best matchesp.server.
-                       // include a timestamp with addresses.
+                       // include a timestamp with addressesp.server.
-       // timeout could fire when we were only half done sending the blocks.
+       // timeout could fire when we were only half done sending the blocksp.server.
-               // The database did not return any further hashes.  Break out of
+               // The database did not return any further hashesp.server.  Break out of
-               // The database did not return any further hashes.  Break out of
+               // The database did not return any further hashesp.server.  Break out of
+       // Do not accept getaddr requests from outbound peersp.server.  This reduces
+       // fingerprinting attacksp.server.
-       // Push the addresses.
+       // Push the addressesp.server.
-       // addresses, and last seen updates.
+       // addresses, and last seen updatesp.server.
```
Looks like there are several find/replace errors.

---
```diff
-func isValidBIP0111(p *peer.Peer, cmd string) bool {
+func isValidBIP0111(p *serverPeer, cmd string) bool {
```

I think the code should be consistent about using `sp` for `*serverPeer`.

---
```diff
+               sp := &serverPeer{
+                       server:          s,
+                       requestedTxns:   make(map[wire.ShaHash]struct{}),
+                       requestedBlocks: make(map[wire.ShaHash]struct{}),
+                       ...
+               }
+               s.peerConfig.Listeners = peer.MessageListeners{
+                       OnVersion:     sp.OnVersion,
+                       ...
+               }
+       sp.Peer = peer.NewOutboundPeer(s.peerConfig, na)
```

Given these is repeated twice, I'd prefer to see functions for these.  Also, I think it would be safer to have a new `peerConfig` for each new peer now since they are encapsulating unique state.  I realize the current code will work because the `peer` package makes a copy of the config, but this feels ripe for a race by only replacing the listeners on a shared `peerConfig` instance with multiple concurrent peers -- epecially when we move to the multi-peer sync code.  I'd suggest something like:

```Go
sp := newServerPeer(s, isPersistent)
sp.Peer = peer.NewOutboundPeer(newPeerConfig(sp), na)
```

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,17 @@
+it provides a mechanism to register and manage custom handlers which will be
+invoked on receiving the given message.
+
+Creating a peer requires a method to retrieve the newest block sha, a nonce, a
```
This is no longer accurate.  Please update it to reflect reality.

tuxcanfly commented:
@davecgh Thanks. Updated the commit to address the items pointed out in the review.

davecgh commented:
Latest updates look good.  I need to go through all of the changes as a whole.  In the mean time, can you manually test all RPC calls that were touched?

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,23 @@
+Inbound peers accept a connection and respond to the version message to begin
+version negotiation.
+
+Outbound peers dial a remote address and use the connection to push the initial
```
This is inaccurate.  Outbound peer do not dial.  The caller is responsible for that.

tuxcanfly replied:
OK, replaced the dial part.

davecgh commented on peer/doc.go:
```diff
@@ -0,0 +1,23 @@
+Outbound peers dial a remote address and use the connection to push the initial
+version message.
+
+Both peers accept a configuration to customize user agent, service flag,
```
I'd add a "such as" here since I don't believe this list covers everything that is configurable.

tuxcanfly replied:
OK, added.

davecgh commented on peer/mruinvmap.go:
```diff
@@ -15,13 +16,17 @@ import (
 // MruInventoryMap provides a map that is limited to a maximum number of items
```
Please call out this is a concurrent safe map (since the changes have made that true).

davecgh commented on peer/mrunoncemap.go:
```diff
@@ -0,0 +1,120 @@
+	"sync"
+)
+
+// mruNonceMap provides a map that is limited to a maximum number of items
```
Same here.

davecgh commented on peer/mrunoncemap.go:
```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2013-2015 The btcsuite developers
```
Please the copyright to just be 2015.  This is new code.

davecgh commented on peer/mrunoncemap_test.go:
```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2013-2015 The btcsuite developers
```
Please the copyright to just be 2015.  This is new code.

davecgh commented:
The copyright for `server.go` needs to be updated to `2013-2015` since this is changing it.

tuxcanfly commented:
@davecgh Thanks, updated the headers and copyright as mentioned.

tuxcanfly commented:
@davecgh Tested `getpeerinfo` on testnet, mainnet and found it to be working as before. Please review.

davecgh commented:
The new `OnGetHeaders` is not ignoring requests before synced like the previous `handleGetHeadersMsg` was.

tuxcanfly commented:
Thanks for the catch, restored it. Looks like it was lost during the initial migration.

davecgh commented:
I believe the `isValidBIP0111` bits should move into the peer package.  The functionality is required of all peers based on the protocol version on the supported services the caller claims.

davecgh commented:
```diff
+       // Add addresses known for this peer.
+       for _, na := range addresses {
+               sp.addKnownAddress(addrmgr.NetAddressKey(na))
+       }
```

This section in `server.go` is wrong.  The `p.PushAddrMsg` function does not necessarily send all of the addresses passed to it as detailed by the comments on the function.  Also, it is really inefficient due to the fact that the new `sp.addKnownAddress` grabs and releases the mutex.  As a result, this would end up having to lock/unlock something like 10000 times (once for every address in the address manager) on every `getaddr` message.

For the first part, I suspect the `p.PushAddrMsg` should return of slice of which addresses were actually pushed and for the second part I suggest creating a `sp.addKnownAddresses(addresses []*wire.NetAddress)` which only locks once, updates the map with all the entries, and then unlocks.

davecgh commented:
Also, I think it would make sense to have a `func (sp *serverPeer) pushAddrMsg(addresses []*wire.NetAddress)` which encapsulates the filtering functionality rather than having to remember to do it multiple times as it is now.

davecgh commented:
A couple of other observations:

* The `type peerState struct {` definition should be grouped with the functions defined on it
* All of the `*serverPeer` functions should be grouped together - in particular, currently `OnRead` and `OnWrite` are defined in the middle of server funcs

tuxcanfly commented:
@davecgh Thanks, I have pushed a commit addressing the issues pointed out:

* moved `isValidBIP0111` to `Peer`
* moved `pushAddrMsg` to `serverPeer`
* grouped related type and funcs in `server.go`

davecgh commented:
I think `PushAddrMsg` should still be in peer, but it should return the created message so the caller can use the actual sent addresses.  The reason is the SPV wallet is going to want to be able to respond to getaddr messages as well and will end up needing the same type of functionality.

davecgh commented:
The `server` type still has a `peerState` member that is now unused and `newServer` is creating it.  Also, when it is created in `peerHandler` it's not adjusting the `maxOutboundPeers` as it should be (and is doing in `newServer` currently).

davecgh commented:
Latest commit seems to fix the remaining two known issues.  I'm going to give it all another pass and do a lot more testing on it.

davecgh commented:
The initial version message push is advertising the services set on the peer instead of the ones configured in the `peer.Config` struct.  This is wrong since it means that it will end advertising the remote peer's service flags instead of the ones supported by the local peer.  Here is the fix:

```diff
diff --git a/peer/peer.go b/peer/peer.go
index c05ee7d..c7b4a4d 100644
--- a/peer/peer.go
+++ b/peer/peer.go
@@ -791,7 +791,7 @@ func (p *Peer) pushVersionMsg() error {
        msg.AddrYou.Services = wire.SFNodeNetwork

        // Advertise the services flag
-       msg.Services = p.Services()
+       msg.Services = p.cfg.Services

        // Advertise our max supported protocol version.
        msg.ProtocolVersion = int32(p.ProtocolVersion())
```

davecgh commented:
Please squash all commits.

davecgh commented:
As a point of reference, all previous comments up until this point have been taken care of.

tuxcanfly commented:
Updated commit message and added a `TODO` regarding `ourNA`.

jrick commented on peer/mrunoncemap_test.go:
```diff
@@ -0,0 +1,152 @@
+
+		// Ensure the limited number of most recent entries in the list
+		// exist.
+		for j := numNonces - 1; j >= numNonces-test.limit; j-- {
```
Can this and the loop below start iterate with increasing j's?  For this one it should start at the limit and iterate up to numNonces, and the one below would iterate from 0 up to the limit. imo that would be more readable than iterating them backwards.

tuxcanfly replied:
OK, updated to iterate in increasing order.

jrick commented:
```
// MessageListeners defines callback function pointers to invoke with message 
// listeners.  Since all of the functions are nil by default, all listeners are 
// effectively ignored until their handlers are set to a concrete callback. 
```

I think this should be reworded as it could be interpreted to mean that these callbacks can be modified through this struct at runtime after a peer is created (it won't race because it creates a copy, but it won't work as intended either).

It also MUST explain how these callbacks are executed.  Per a single peer, do these callbacks execute concurrently with multiple messages or are they dispatched serially?

jrick commented on peer/peer_test.go:
```diff
@@ -0,0 +1,640 @@
+		return
+	}
+
+	// TODO: actually test the following methods
```
bump

jrick commented on peer/peer_test.go:
```diff
@@ -0,0 +1,640 @@
+		wantConnected:       true,
+		wantVersionKnown:    true,
+		wantVerAckReceived:  true,
+		wantLastPingTime:    *new(time.Time),
```
time.Time{}?

davecgh replied:
Agree

tuxcanfly replied:
OK, updated.

davecgh commented:
For reference, here is an overview of the major testing points I've done:

* Ensure general communications work as before including all supported p2p messages are handled properly (version negotiation, transactions, blocks, inventory, data retrieval, addresses, mempool,  bloom filters)
* Manually verifying properly generated version messages via tracnig
* Maximum number of peers is respected
* Eight outbound peers are maintained
* Manual testing against a couple of SPV wallets connecting to it
* Check output of peer data in various RPC commands (results below)
* Adding permanent and temp peers via `node` and `addnode` commands
* Disconnecting peers via `node` command

Connection count from `getinfo`:
> $ btcctl getinfo | grep connections
> "connections": 15,

Outbound and inbound peers shown properly in `getpeerinfo`:
> $ btcctl getpeerinfo | grep "\"inbound\": false" | wc -l
> 8
> $ btcctl getpeerinfo | grep "\"inbound\": true" | wc -l
> 7


Total bytes read and written from `getnettotals`:
> $ btcctl getnettotals
> {
>   "totalbytesrecv": 402747331,
>   "totalbytessent": 3840105502,
>   ...
> }

davecgh commented:
Once @jrick's comments are addressed, this is ready for master.

tuxcanfly commented:
@davecgh Addressed comments by @jrick, rebased and squashed.

davecgh commented:
Everything has been addressed for me.

OK

jrick commented:
ok
